File: Playlist/StatusImageTransformer.m
//
//  StatusImageTransformer.m
//  Cog
//
//  Created by Vincent Spader on 2/21/08.
//  Copyright 2008 __MyCompanyName__. All rights reserved.
//

#import "StatusImageTransformer.h"
#import "PlaylistEntry.h"


@implementation StatusImageTransformer

@synthesize playImage;
@synthesize queueImage;
@synthesize errorImage;
@synthesize stopAfterImage;

+ (Class)transformedValueClass { return [NSImage class]; }
+ (BOOL)allowsReverseTransformation { return NO; }

- (id)init
{
	self = [super init];
	if (self)
	{
		self.playImage = [NSImage imageNamed:@"playTemplate"];
		self.queueImage = [NSImage imageNamed:@"NSAddTemplate"];
		self.errorImage = [NSImage imageNamed:@"NSStopProgressTemplate"];
		self.stopAfterImage = [NSImage imageNamed:@"stopTemplate"];
	}
	
	return self;
}

// Convert from string to RepeatMode
- (id)transformedValue:(id)value {
    if (value == nil) return nil;

	if ([value isEqualToString:@"playing"])
	{
		return self.playImage;
	}
	else if ([value isEqualToString:@"queued"])
	{
		return self.queueImage;
	}
	else if ([value isEqualToString:@"error"]) {
		return self.errorImage;
	}
	else if ([value isEqualToString:@"stopAfter"]) {
		return self.stopAfterImage;
	}

	return nil;
}


@end

File: Playlist/StatusImageTransformer.swift
import Foundation
import AppKit

@objc(StatusImageTransformer)
class StatusImageTransformer: ValueTransformer {
    var playImage: NSImage?
    var queueImage: NSImage?
    var errorImage: NSImage?
    var stopAfterImage: NSImage?

    override class func transformedValueClass() -> AnyClass {
        return NSImage.self
    }

    override class func allowsReverseTransformation() -> Bool {
        return false
    }

    override init() {
        super.init()
        if #available(macOS 11.0, *) {
            self.playImage = NSImage(systemSymbolName: "play.fill", accessibilityDescription: NSLocalizedString("PlayingTrackTooltip", comment: ""))
            self.queueImage = NSImage(systemSymbolName: "plus", accessibilityDescription: NSLocalizedString("QueuedTrackTooltip", comment: ""))
            self.errorImage = NSImage(systemSymbolName: "nosign", accessibilityDescription: NSLocalizedString("ErrorTrackTooltip", comment: ""))
            self.stopAfterImage = NSImage(systemSymbolName: "stop.fill", accessibilityDescription: NSLocalizedString("StopAfterTrackTooltip", comment: ""))
        } else {
            self.playImage = NSImage(named: "playTemplate")
            self.queueImage = NSImage(named: "NSAddTemplate")
            self.errorImage = NSImage(named: "NSStopProgressTemplate")
            self.stopAfterImage = NSImage(named: "stopTemplate")
        }
    }

    // Convert from string to image
    override func transformedValue(_ value: Any?) -> Any? {
        if value == nil {
            return nil
        }

        if let stringValue = value as? String {
            switch stringValue {
            case "playing":
                return self.playImage
            case "queued":
                return self.queueImage
            case "error":
                return self.errorImage
            case "stopAfter":
                return self.stopAfterImage
            default:
                return nil
            }
        }

        return nil
    }
}
